Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/ExposureKeyHistoryCalculations.kt

```diff
@@ -1,39 +1,72 @@
 package de.rki.coronawarnapp.submission
 
+import KeyExportFormat
+import androidx.annotation.VisibleForTesting
 import com.google.android.gms.nearby.exposurenotification.TemporaryExposureKey
+import org.joda.time.Days
+import org.joda.time.Instant
 
 class ExposureKeyHistoryCalculations(
     private val transmissionRiskVectorDeterminator: TransmissionRiskVectorDeterminator,
+    private val daysSinceOnsetOfSymptomsVectorDeterminator: DaysSinceOnsetOfSymptomsVectorDeterminator,
     private val keyConverter: KeyConverter
 ) {
 
+    companion object {
+        const val MAX_DAYS_IN_PAST = 14
+        const val VECTOR_LENGTH = 15
+        const val TEN_MINUTES_IN_MILLIS = (10 * 60 * 1000).toLong()
+    }
+
     fun transformToKeyHistoryInExternalFormat(
         keys: List<TemporaryExposureKey>,
         symptoms: Symptoms
     ) =
         toExternalFormat(
-            toSortedHistory(limitKeyCount(keys)),
-            transmissionRiskVectorDeterminator.determine(symptoms)
+            limitKeyCount(toSortedHistory(keys)),
+            transmissionRiskVectorDeterminator.determine(symptoms),
+            daysSinceOnsetOfSymptomsVectorDeterminator.determine(symptoms, VECTOR_LENGTH)
         )
 
-    fun <T> limitKeyCount(keys: List<T>): List<T> =
-        keys.take(MAXIMUM_KEYS)
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun limitKeyCount(keys: List<TemporaryExposureKey>, now: Instant = Instant()) =
+        keys.filter { it.daysAgo(now) <= MAX_DAYS_IN_PAST }
 
-    fun toExternalFormat(
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun toExternalFormat(
         keys: List<TemporaryExposureKey>,
-        transmissionRiskVector: TransmissionRiskVector
-    ) =
-        keys.mapIndexed { index, key ->
-            // The latest key we receive is from yesterday (i.e. 1 day ago),
-            // thus we need use index+1
-            keyConverter.toExternalFormat(key, transmissionRiskVector.getRiskValue(index + 1))
+        transmissionRiskVector: TransmissionRiskVector,
+        daysSinceOnsetOfSymptomsVector: DaysSinceOnsetOfSymptomsVector,
+        now: Instant = Instant()
+    ): List<KeyExportFormat.TemporaryExposureKey> {
+        val result = mutableListOf<KeyExportFormat.TemporaryExposureKey>()
+        keys.groupBy { it.daysAgo(now) }.forEach { entry ->
+            val daysAgo = entry.key
+            entry.value.forEach {
+                result.add(
+                    keyConverter.toExternalFormat(
+                        it,
+                        transmissionRiskVector.getRiskValue(daysAgo),
+                        daysSinceOnsetOfSymptomsVector[daysAgo]
+                    )
+                )
+            }
         }
+        return result.toList()
+    }
 
-    fun toSortedHistory(keys: List<TemporaryExposureKey>) =
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun toSortedHistory(keys: List<TemporaryExposureKey>) =
```

Comment:
Proposal: I would prefer if we added a bit more semantics here to make it easier to understand what's the result. I think the set is sorted with the the most recent key(s) first, right? If that's the case, maybe rename it to `sortWithRecentKeyFirst`.